File: 백준/Gold/2470. 두 용액/두 용액.py
# 2470 두 용액
import sys

N = int(sys.stdin.readline())

prob = sorted(list(map(int, sys.stdin.readline().split())))


def binary_search():
    now_answer = []
    min_value = float('inf')
    for i in range(N):
        left = i + 1
        right = N - 1
        middle = 0
        while left <= right:
            middle = (left + right) // 2
            if prob[middle] + prob[i] == 0:
                now_answer = sorted([prob[middle], prob[i]])
                print(" ".join(map(str, now_answer)))
                return
            elif prob[middle] + prob[i] < 0:
                left = middle + 1
            else:
                right = middle - 1

            if abs(min_value) > abs(prob[middle] + prob[i]):
                min_value = abs(prob[middle] + prob[i])
                now_answer = [prob[middle], prob[i]]

    now_answer = sorted(now_answer)
    print(" ".join(map(str, now_answer)))


binary_search()

File: 백준/Gold/2470. 두 용액/두 용액.java
//  두 용액

import java.util.*;
import java.lang.*;
import java.io.*;

import static java.lang.Math.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        int[] prob = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .sorted().toArray();

        int left = 0;
        int right = N - 1;
        int answer = prob[left] + prob[right];
        int leftAnswer = prob[left];
        int rightAnswer = prob[right];

        while (left < right) {
            int now = prob[left] + prob[right];

            if (abs(answer) > abs(now)) {
                answer = now;
                leftAnswer = prob[left];
                rightAnswer = prob[right];
            }

            if (now < 0) {
                left++;
            } else if (now > 0) {
                right--;
            } else {
                System.out.printf("%d %d", prob[left], prob[right]);
                return;
            }

        }

        System.out.printf("%d %d", leftAnswer, rightAnswer);
    }
}
